clients/src/main/java/org/apache/kafka/common/requests/ApiError.java

```diff
@@ -36,10 +37,16 @@
     private final String message;
 
     public static ApiError fromThrowable(Throwable t) {
+        Throwable throwableToBeEncode = t;
+        // Future will wrap the original exception with completionException, which will return unexpected UNKNOWN_SERVER_ERROR.
+        if (t instanceof CompletionException) {
+            throwableToBeEncode = t.getCause();
```

Comment:
fix 1: unwrap the CompletionException to get the original exception inside.

Reply:
fix 1: unwrap the `CompletionException` to get the original exception inside. Even if we don't want the `NotControllerException` return back to client, we need to know it to do some check.